Changed from setup.py to pyproject.toml

MitchiLaser opened the pull request:
The whole configuration has been moved from setup.py to the pyproject.toml file because setuptools now recommends it that way.

MitchiLaser committed:
auto copyright update when generating documentation

MitchiLaser committed:
conf.py -> format string required

MitchiLaser committed:
Merge branch 'PhiLFitters:master' into master

MitchiLaser committed:
moved completely from setup.py to pyproject.toml

MitchiLaser commented:
After some discussions with Günter and Achim M. the value for `requires-python` should be set to 3.6 because all the computers of the students computer pool room have python version 3.6 installed (currently the newest for OpenSUSE)

MitchiLaser committed:
Merge branch 'PhiLFitters:master' into master

MitchiLaser committed:
pyproject.toml updated value for requires-python

after a small discussion the conclusion was to set this value to python version 3.6

cverstege commented on pyproject.toml:
```diff
@@ -0,0 +1,61 @@
+name = "kafe2"
+description = "Karlsruhe Fit Environment 2: a package for fitting and elementary data analysis"
+authors = [
+	{ name="Daniel Savoiu", email="daniel.savoiu@cern.ch" },
```
While you're at it...
You could add the following authors:
```
{name="Johannes Gäßler", email="johannesg@5d6.de"},
{name="Cedric Verstege", email="cedric.verstege@cern.ch"},
{name="Günter Quast, email="G.Quast@kit.edu"},
```
I hope @JohannesGaessler and @GuenterQuast are fine with this as well.

MitchiLaser replied:
Thanks for the notice, I've added them to the file

MitchiLaser committed:
added aothors in pyproject.toml

cverstege commented:
I will test this setup in the coming days.

Note on merging: I'd prefer a squash and merge, to squash the changes into a single commit.

cverstege commented:
The `pyproject.toml` is working with setuptools. Using the wheel it created, kafe2go is also present. I didn't encouter any issues. Additionally all unit tests are looking good, when installing kafe via the wheel build by setuptools using the `pyproject.toml` file.
However, the GitHub workflows need to be updated.

MitchiLaser commented:
The following workflow can be adopted for this repository to make it work again: [https://github.com/MitchiLaser/python-feiertage/blob/main/.github/workflows/release_to_pypi.yml](https://github.com/MitchiLaser/python-feiertage/blob/main/.github/workflows/release_to_pypi.yml)

I will take a closer Look at this today in the evening

cverstege commented:
> The following workflow can be adopted for this repository to make it work again: https://github.com/MitchiLaser/python-feiertage/blob/main/.github/workflows/release_to_pypi.yml
> 
> I will take a closer Look at this today in the evening

Here's a diff I just created:
```
diff --git a/.github/workflows/pythonpublish.yml b/.github/workflows/pythonpublish.yml
index d80c7c0..19491a4 100644
--- a/.github/workflows/pythonpublish.yml
+++ b/.github/workflows/pythonpublish.yml
@@ -21,11 +21,11 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install setuptools wheel twine
-    - name: Build and publish
+        pip install build twine
+    - name: Build
+      run: python -m build
+    - name: Publish to PyPi
       env:
         TWINE_USERNAME: ${{ secrets.PYPI_USERNAME }}
         TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
-      run: |
-        python setup.py sdist bdist_wheel
-        twine upload dist/*
+      run: twine upload dist/*
diff --git a/.github/workflows/pythonpublish_manual.yml b/.github/workflows/pythonpublish_manual.yml
index 41db3d3..51bc1fb 100644
--- a/.github/workflows/pythonpublish_manual.yml
+++ b/.github/workflows/pythonpublish_manual.yml
@@ -19,11 +19,12 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install setuptools wheel twine
-    - name: Build and publish
+        pip install build twine
+    - name: Build
+      run: python -m build
+    - name: Publish to PyPi
       env:
         TWINE_USERNAME: ${{ secrets.PYPI_USERNAME }}
         TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
-      run: |
-        python setup.py sdist bdist_wheel
-        twine upload --verbose dist/*
+      run: twine upload --verbose dist/*
```

cverstege commented on pyproject.toml:
```diff
@@ -0,0 +1,64 @@
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
+]
+#TODO dependency versions
+dependencies = [
+        "iminuit",
+        "NumPy",
+        "Numdifftools",
+        "Scipy",
+        "tabulate",
+        "matplotlib",
+        "PyYaml",
+        "six",
+        "funcsigs",
+        "SymPy",
+]
```
The dev dependencies from [dev_dependencies.txt](https://github.com/PhiLFitters/kafe2/blob/master/dev_dependencies.txt) can be added using this code template:
```
[project.optional-dependencies]
dev = [
    "pylint ~=2.14.0",
    "toml ~=0.10.2",
    "yapf ~=0.32.0",
]
```

cverstege replied:
I'm not 100% sure this works. I did not test this yet.

cverstege replied:
The dev dependencies are:
```
numpy
scipy
iminuit
matplotlib
numdifftools
pyyaml
six
funcsigs
tabulate
coverage
sphinx
sphinx_bootstrap_theme
mock
sympy
```
And can probably be added using the syntax above. I think the dependencies already given as a default can be excluded from this list.
When they are added inside the pyrproject.toml the dev_dependencies.txt file can be deleted as well.

cverstege commented on pyproject.toml:
```diff
@@ -0,0 +1,64 @@
+dynamic = ["version"]
+
+[tool.setuptools.dynamic]
+version = {attr = "kafe2._version_info.__version__"}	# todo: check wether this works
```
Seems to work when building on my machine ;)

cverstege committed:
Remove travis.yml, moved to github actions

cverstege committed:
.gitignore add dist folder from builds

MitchiLaser committed:
Added optional dependencies and removed TODO comment from
pyproject.toml, completely rewritten github workflows

MitchiLaser commented:
I made some changes to the Github CD Workflows.

Unfortunately I am not able to test them. Maybe these could be testes using PyPI test server. This can be done by adding the arguments ` --repository testpypi` to the call of `twine`. More information about the PyPI testing environment can be read here:
[https://packaging.python.org/en/latest/guides/using-testpypi/](https://packaging.python.org/en/latest/guides/using-testpypi/)
Maybe for the test an account has to be registerd at test.pypi.org (this might be useful in the future for nightly releases), either with the same username and password as the current account which is being used for the kafe2 distribution or with a new one (Attention: The login informations have to be added to the github secrets).

Accorgind to github I screw up with the commit description in the last commit but from my point of view this should not be a big deal because this will be automatically replaced with a new one after merging to the master branch. Sorry for the inconvenience.

Hopefully now the new build system is fully integrated and can be used to completely replace the old one :)

cverstege commented on pyproject.toml:
```diff
@@ -0,0 +1,71 @@
+    "pylint ~=2.14.0",
+    "toml ~=0.10.2",
+    "yapf ~=0.32.0",
+]
```
My previous comment was just some boilerplate. The actual dev dependencies can be found in https://github.com/PhiLFitters/kafe2/blob/master/dev_dependencies.txt
Sorry for the confusion

cverstege commented on .github/workflows/pythonpublish_manual.yml:
```diff
@@ -5,25 +5,32 @@ name: Upload Python Package (manual)
-      run: |
-        python setup.py sdist bdist_wheel
-        twine upload --verbose dist/*
+      run: twine upload dist/*
```
I think the --verbose flag should be present in the manually triggered workflow, as before.

MitchiLaser committed:
removed optional dependencies from pyproject.toml and fixed indentation problem in workflow files

MitchiLaser commented:
Sorry, i forgot about the `---verbose` flag.

MitchiLaser commented:
There is still a problem in the workflow files (maybe an indentation error). I will take a closer look at this tomorrow (currently sitting in a train and the internet connection is really poor).

JohannesGaessler committed:
Update Johannes Gäßler email address

MitchiLaser committed:
fixed indentation problem in workflow file

JohannesGaessler commented on pyproject.toml:
```diff
@@ -0,0 +1,64 @@
+description = "Karlsruhe Fit Environment 2: a package for fitting and elementary data analysis"
+authors = [
+	{ name="Daniel Savoiu", email="daniel.savoiu@cern.ch" },
+	{ name="Johannes Gäßler", email="johannesg@5d6.de" },
```
```suggestion
	{ name="Johannes Gäßler", email="johannes.gaessler@cern.ch" },
```

MitchiLaser replied:
This was already changed

JohannesGaessler commented on pyproject.toml:
```diff
@@ -0,0 +1,64 @@
+        "Topic :: Scientific/Engineering :: Visualization",
+        "License :: OSI Approved :: GNU General Public License v3 (GPLv3)",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.8",
```
```suggestion
        "Programming Language :: Python :: 3",
        "Programming Language :: Python :: 3.6",
        "Programming Language :: Python :: 3.7",
        "Programming Language :: Python :: 3.8",
```

MitchiLaser committed:
Update Johannes E-Mail address

AlexanderKaschta committed:
FIX: Remove illegal characters from file names for Windows

MitchiLaser committed:
integrated dev depepdencies into pyproject.toml

MitchiLaser commented:
I'm sorry because it took a while yesterday (and the train ride was far from comfortable). Now all the indentation problems seem to be fixed and the workflows should run. Furthermore I updated the E-Mail Address from @JohannesGaessler in the pyproject.toml file and added all the dev dependencies. The file dev_dependencies.txt was deleted because it is not longer needed any more.

JohannesGaessler commented on .github/workflows/pythonpublish.yml:
```diff
@@ -9,23 +9,29 @@ on:
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.x'
```
Which Python version does this use? The latest one?

MitchiLaser replied:
Yes, it is the latest stable version
[https://github.com/actions/setup-python/blob/0d5da6a89a3aa5b4dbd48e38b0c9d678b0caa0e5/docs/advanced-usage.md?plain=1#L103](https://github.com/actions/setup-python/blob/0d5da6a89a3aa5b4dbd48e38b0c9d678b0caa0e5/docs/advanced-usage.md?plain=1#L103)

JohannesGaessler commented on Makefile:
```diff
@@ -0,0 +1,27 @@
+	python -m build
+
+clean:
+	rm -rf build/
+	rm -rf dist/
+	rm -rf kafe2.egg-info/
+	rm -rf venv/
+	rm -rf `find . -type d -name __pycache__`
```
```suggestion
	-rm -rf build/
	-rm -rf dist/
	-rm -rf kafe2.egg-info/
	-rm -rf venv/
	-rm -rf `find . -type d -name __pycache__`
```

I think `rm` returns a non-zero exit code if the files do not exist. This will prevent further commands in the `clean` target from being executed. You can suppress errors by prefixing `-`.

MitchiLaser replied:
This is not necessary since the target `clean` is listed in the `.IGNORE` Rule, see [https://www.gnu.org/software/make/manual/html_node/Special-Targets.html](https://www.gnu.org/software/make/manual/html_node/Special-Targets.html) for further information

JohannesGaessler commented on pyproject.toml:
```diff
@@ -0,0 +1,68 @@
+version = {attr = "kafe2._version_info.__version__"}
+
+[project.urls]
+"Homepage" = "https://github.com/PhiLFitters/kafe2"
```
```suggestion
"Homepage" = "https://philfitters.github.io/kafe2/"
```

MitchiLaser replied:
Thanks. I wasn't aware of the fact that this website even existed

cverstege commented on pyproject.toml:
```diff
@@ -43,6 +43,10 @@ dependencies = [
+        "coverage",
+        "sphinx",
+        "sphinx_bootstrap_theme",
+        "mock",
```
You've added the dev dependencies to the normal dependencies, which is not needed.
Please add them as optional dependencies with:
```
[project.optional-dependencies]
dev = [
    "coverage",
    "sphinx",
    "sphinx_bootstrap_theme",
    "mock",
    "pytest",
    "flake8"
]

MitchiLaser replied:
OK. I hope i finally understood what you want to do with the optional dependencies. In this case calling `pip install kafe2[dev]` will install kafe2 and the optional dependencies listed in the `dev` array.

cverstege replied:
I have to see if this works in editable mode as well. But yes, this is what I meant :D

MitchiLaser replied:
I already tested it on my machine by setting up a virtual environment and calling `pip install dist/kafe2-2.8.0.tar.gz[dev]` in there and all the optional dependencies were also installed by pip.

cverstege replied:
I just tested this in editable mode as well. When in the kafe2 project dir `pip install -e .[dev]` installs kafe2 in editable mode including all requirements. That's ideal for development.

MitchiLaser committed:
Update dependencies and python versions in pyproject.toml

MitchiLaser commented:
Most of the suggested changes have been added, the Makefile was left untouched

cverstege commented on Makefile:
```diff
@@ -0,0 +1,27 @@
+upload: build
+	twine upload dist/*
+
+venv:	build #docs
+	python -m venv venv/
+	. venv/bin/activate; pip install dist/kafe2*.tar.gz
+	. venv/bin/activate; pip install ptpython
```
We could add a development venv like this:
```
python -m venv venv/
. venv/bin/activate; pip install -e .[dev]
```

MitchiLaser replied:
Yes, this is a todo we already thought about.

cverstege commented:
Could you please rebase onto the dev branch? Then I'll merge it.

MitchiLaser commented:
Done?

MitchiLaser commented:
:D
